core/src/OpenACC/Kokkos_OpenACC_Atomic.hpp

```diff
@@ -0,0 +1,1964 @@
+#else
+  unsigned long long int *ptr = const_cast<unsigned long long int *>(dest);
+  #pragma acc atomic capture
+  { retval = *ptr; *ptr = val; }
+#endif
+  return retval;
+}
+
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+#pragma acc routine seq
+template <>
+float atomic_exchange(
+    volatile float *const dest,
+    const float &val) {
+  float retval;
+  retval = atomicExch((float *)dest, val);
+  return retval;
+}
+#endif
+
+// atomic_compare_exchange -----------------------------------------------------
+
+#pragma acc routine seq
+template <typename T>
+__inline__  T atomic_compare_exchange(
+    volatile T *const dest, const T &compare,
+    const T &val) {
+	if( acc_on_device(acc_device_not_host) ) {
+		#warning atomic_compare_exchange(T) not correctly implemented in OpenACC 
+	}
+    T retval;
+	T *ptr = const_cast<T *>(dest);
+    {
+      retval = *ptr;
+      if( retval == compare) *ptr = val;
+    }
+    return retval;
+}
+
+#pragma acc routine seq
+template <>
+__inline__  int atomic_compare_exchange(
+    volatile int *const dest, const int &compare,
+    const int &val) {
+    int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicCAS((int *)dest, compare, val);
+#else
+	int *ptr = const_cast<int *>(dest);
+#warning atomic_compare_exchange(int) not correctly implemented in OpenACC 
+    {
+      retval = *ptr;
+      if( retval == compare) *ptr = val;
+    }
+#endif
+    return retval;
+}
+
+#pragma acc routine seq
+template <>
+__inline__  unsigned int atomic_compare_exchange(
+    volatile unsigned int *const dest, const unsigned int &compare,
+    const unsigned int &val) {
+    unsigned int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicCAS((unsigned int *)dest, compare, val);
+#else
+	unsigned int *ptr = const_cast<unsigned int *>(dest);
+#warning atomic_compare_exchange(unsigned int) not correctly implemented in OpenACC 
+    {
+      retval = *ptr;
+      if( retval == compare) *ptr = val;
+    }
+#endif
+    return retval;
+}
+
+#pragma acc routine seq
+template <>
+__inline__  unsigned long long int atomic_compare_exchange(
+    volatile unsigned long long int *const dest, const unsigned long long int &compare,
+    const unsigned long long int &val) {
+    unsigned long long int retval;
+#ifdef KOKKOS_ENABLE_CUDA_ATOMIC_INTRINSICS
+	retval = atomicCAS((unsigned long long int *)dest, compare, val);
+#else
+	unsigned long long int *ptr = const_cast<unsigned long long int *>(dest);
+#warning atomic_compare_exchange(unsigned long long int) not correctly implemented in OpenACC 
+    {
+      retval = *ptr;
+      if( retval == compare) *ptr = val;
+    }
+#endif
+    return retval;
+}
+
+// atomic_assign ---------------------------------------------------------------
+
+#pragma acc routine seq
+template <typename T>
+__inline__ void atomic_assign(
+    volatile T *const dest,
+    const T & val) {
+	T *ptr = const_cast<T *>(dest);
+//#pragma acc atomic write
```

Comment:
That code was for unsupported atomic data types; I changed the code so that for unsupported atomic data types, it will abort if running on a device but it will sequentially execute the construct (without atomicity) if running on a host.